### PR TITLE
Add Datadog profiler

### PR DESCRIPTION
**Pull request by daxmc99**
- Add Datadog as a supported profiler
~~- Print an error if GOMAXPROCs fails~~

Adds Datadog profiler  
Part of https://github.com/sourcegraph/zoekt/issues/294

### CODE REVIEW

File: cmd/zoekt-archive-index/main.go [hunk 1]
```diff
@@ -210,8 +210,10 @@ func main() {
 	flag.Parse()
 
 	// Tune GOMAXPROCS to match Linux container CPU quota.
-	_, _ = maxprocs.Set()
-
+	_, err := maxprocs.Set()
+	if err != nil {
+		log.Printf("could not set GOMAXPROCS: %v\n", err)
+	}
```
Comment (keegancsmith) [inline cmd/zoekt-archive-index/main.go]: can you get rid of this change? It results in log spam. We don't only run in containers and linux (eg macbooks)

File: cmd/zoekt-sourcegraph-indexserver/main.go [hunk 2]
```diff
@@ -125,6 +125,8 @@ const (
 	indexStateEmpty       indexState = "empty"        // index is empty (empty repo)
 )
 
+const ServiceName = "zoekt-sourcegraph-indexserver"
```
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/main.go]: this doesn't need to be exported. Additionally this is only read in one place, just inline the string there rather than having indirection to the value.

File: cmd/zoekt-sourcegraph-indexserver/main.go [hunk 3]
```diff
@@ -817,12 +800,19 @@ func newServer(conf rootConfig) (*Server, error) {
...
+	if err != nil {
+		log.Printf("could not set GOMAXPROCS: %v\n", err)
+	}
 
 	// Set the sampling rate of Go's block profiler: https://github.com/DataDog/go-profiler-notes/blob/main/guide/README.md#block-profiler.
-	// The block profiler is disabled by default.
+	// The block profiler is disabled by default and should be enabled with care in production
 	runtime.SetBlockProfileRate(conf.blockProfileRate)
 
+	err = profiler.Init(ServiceName, zoekt.Version, conf.blockProfileRate)
```
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/main.go]: why did you move profiler.Init to here? Previously it happened in startServer. I believe that was intentional so we didn't capture short lived debug invocations of this binary.

File: cmd/zoekt-webserver/main.go [hunk 4]
```diff
@@ -150,7 +152,10 @@ func main() {
 	}
 
 	initializeJaeger()
-	initializeGoogleCloudProfiler()
+	err := profiler.Init(ServiceName, zoekt.Version, -1)
+	if err != nil {
+		log.Printf("could not initialize profiler: %s", err.Error())
```
Comment (keegancsmith) [inline cmd/zoekt-webserver/main.go]: if we only ever log the error, can we improve the ergonomics here and just log internally in profiler.Init and not return an error.

File: internal/profiler/profiler.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+// Init starts the supported profilers IFF the environment variable is set.
+func Init(svcName, version string, blockProfileRate int) error {
+	if os.Getenv("DD_ENV") != "" {
+		profileTypes := []ddprofiler.ProfileType{ddprofiler.CPUProfile, ddprofiler.HeapProfile}
+		// additional profilers have a performance impact and should be enabled with care
+		if os.Getenv("DD_PROFILE_ALL") != "" || blockProfileRate > 0 {
+			profileTypes = append(profileTypes, ddprofiler.MutexProfile, ddprofiler.BlockProfile)
+		}
+		return ddprofiler.Start(
```
Comment (keegancsmith) [inline internal/profiler/profiler.go]: you are doing an early return here, and then won't start google cloud profiler.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
Nice. I have a few inline comments. I only comment in one place the feedback applies, but most of my inline comments apply to other files the pattern was applied.

**Comment by keegancsmith** [comment]
Test failure looks unrelated. cc @ggilmore 

```
--- FAIL: TestDeltaShards (0.63s)
    --- FAIL: TestDeltaShards/tombstones_affect_document_across_branches (0.12s)
Error:         e2e_test.go:753: step "tombstone foo": finishing builder: repository metadata in shard "/tmp/TestDeltaShardstombstones_affect_document_across_branches1240476489/001/repository_v16.00000.zoekt" contains a different set of branch names than what was requested, which is unsupported in a delta shard build. old: [{Name:main Version:} {Name:release Version:}], new: [{Name:release Version:} {Name:main Version:}]
```

File: cmd/zoekt-sourcegraph-indexserver/main.go (hunk 3 above)
Comment (daxmc99) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: My thought here was to keep this close to the `runtime.SetBlockProfile`. If all the Datadog profiler is enabled with `DD_MAX_PROFILER` then it will set this value. 

I suppose I don't fully understand _why_ we are setting the BlockProfileRate here. I'm lacking the context of why this would be set when not profiling or what affect this value would have before the change. 

As far as capturing short-lived invocations, I'm not sure I follow. In all debug scenarios, I think you will need to setup an external profiler for profiles to be collected and analyzed.

File: cmd/zoekt-archive-index/main.go (hunk 1 above)
Comment (daxmc99) [reply cmd/zoekt-archive-index/main.go]: Reverted! This was due to a (likely unfounded) concern about cgroups v2 with newer version of `containerd` 
See https://github.com/uber-go/automaxprocs/pull/44

File: internal/profiler/profiler.go (hunk 5 above)
Comment (daxmc99) [reply internal/profiler/profiler.go]: I think we want only one profiler enabled at a time

File: cmd/zoekt-sourcegraph-indexserver/main.go (hunk 2 above)
Comment (daxmc99) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: Inlined!

File: cmd/zoekt-sourcegraph-indexserver/main.go (hunk 3 above)
Comment (keegancsmith) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: zoekt-sourcegraph-indexserver can be run not as a daemon, but as a short lived command just for debug purposes. Those do not need datadog/cloud profiler and shouldn't have them.

I think you should move it to where we had it before. runtime.SetBlockProfile doesn't start the block profiler, that is why it was done where ti was before.

File: internal/profiler/profiler.go (hunk 5 above)
Comment (keegancsmith) [reply internal/profiler/profiler.go]: Why? If that is what an operator wants, then they can only enable one envvar?

File: cmd/zoekt-sourcegraph-indexserver/main.go [hunk 6]
```diff
@@ -820,9 +801,10 @@ func newServer(conf rootConfig) (*Server, error) {
 	_, _ = maxprocs.Set()
 
 	// Set the sampling rate of Go's block profiler: https://github.com/DataDog/go-profiler-notes/blob/main/guide/README.md#block-profiler.
-	// The block profiler is disabled by default.
+	// The block profiler is disabled by default and should be enabled with care in production
 	runtime.SetBlockProfileRate(conf.blockProfileRate)
 
+	profiler.Init("zoekt-sourcegraph-indexserver", zoekt.Version, conf.blockProfileRate)
```
Comment (keegancsmith) [inline cmd/zoekt-sourcegraph-indexserver/main.go]: can you just make it so that we call runtime.SetBlockProfileRate inside of profiler.Init? That way webserver also gets to benefit from it.

cc @ggilmore do we actually use this anymore? Should we just get rid of it?

File: internal/profiler/profiler.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+		profileTypes := []ddprofiler.ProfileType{ddprofiler.CPUProfile, ddprofiler.HeapProfile}
+		// additional profilers have a performance impact and should be enabled with care
+		if os.Getenv("DD_PROFILE_ALL") != "" || blockProfileRate > 0 {
+			profileTypes = append(profileTypes, ddprofiler.MutexProfile, ddprofiler.BlockProfile)
+		}
+		err := ddprofiler.Start(
+			ddprofiler.WithService(svcName),
+			ddprofiler.WithVersion(version),
+			ddprofiler.WithProfileTypes(profileTypes...,
+			),
```
Comment (keegancsmith) [inline internal/profiler/profiler.go]:
```suggestion
			ddprofiler.WithProfileTypes(profileTypes...),
```

**Review by keegancsmith — APPROVED** [review]

File: cmd/zoekt-sourcegraph-indexserver/main.go (hunk 6 above)
Comment (ggilmore) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: Are you talking about the block profiler itself? We don't run it continuously, but we have the machinery to enable it by setting an environment variable (this would we be useful while we're debugging something).

**Comment by ggilmore** [comment]
> Test failure looks unrelated. cc @ggilmore
> 
> ```
> --- FAIL: TestDeltaShards (0.63s)
>     --- FAIL: TestDeltaShards/tombstones_affect_document_across_branches (0.12s)
> Error:         e2e_test.go:753: step "tombstone foo": finishing builder: repository metadata in shard "/tmp/TestDeltaShardstombstones_affect_document_across_branches1240476489/001/repository_v16.00000.zoekt" contains a different set of branch names than what was requested, which is unsupported in a delta shard build. old: [{Name:main Version:} {Name:release Version:}], new: [{Name:release Version:} {Name:main Version:}]
> ```

This should be fixed if you merge in master.
